From 1b52ab98e6c54d4794a78b6fc7a2f52071317ead Mon Sep 17 00:00:00 2001
From: Ye He <ye.he@amlogic.com>
Date: Sat, 10 Oct 2026 15:11:51 +0800
Subject: [PATCH] bearer: Cancel disconn_timer upon btd_bearer_disconnected

If a fast reconnect occurs after a rapid disconnect,
the 2-seconds bearer->disconn_timer may still fire and
tear down the newly established link.

Fix this by cancelling the disconn_timer inside
btd_bearer_disconnected().
---
 src/bearer.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/bearer.c b/src/bearer.c
index 25224dc..5544c2f 100644
--- a/src/bearer.c
+++ b/src/bearer.c
@@ -464,6 +464,12 @@ void btd_bearer_disconnected(struct btd_bearer *bearer, uint8_t reason)
 	if (!bearer || !bearer->path)
 		return;
 
+	/* cancel disconnect timer if ACL link has already dropped */
+	if (bearer->disconn_timer) {
+		timeout_remove(bearer->disconn_timer);
+		bearer->disconn_timer = 0;
+	}
+
 	if (!btd_device_is_connected(bearer->device))
 		device_disconnect_watches_callback(bearer->device);
 
-- 
2.53.0

